pkg/cloudprovider/provider/openstack/provider.go

```diff
@@ -87,7 +108,124 @@ func getClient(c *Config) (*gophercloud.ProviderClient, error) {
 		TokenID:          c.TokenID,
 	}
 
-	return openstack.AuthenticatedClient(opts)
+	return goopenstack.AuthenticatedClient(opts)
+}
+
+func (p *provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, bool, error) {
+	var changed bool
+
+	c, _, err := getConfig(spec.ProviderConfig)
+	if err != nil {
+		return spec, changed, fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	client, err := getClient(c)
+	if err != nil {
+		return spec, changed, fmt.Errorf("failed to get a openstack client: %v", err)
+	}
+
+	if c.Region == "" {
+		glog.V(4).Infof("Trying to default region...")
+		regions, err := getRegions(client)
+		if err != nil {
+			return spec, changed, fmt.Errorf("Failed to get regions: %s", err)
+		}
+		if len(regions) == 1 {
+			glog.V(4).Infof("Defaulted region to '%s'", regions[0].ID)
+			changed = true
+			c.Region = regions[0].ID
+		} else {
+			return spec, changed, fmt.Errorf("Could not default region because got '%v' results!", len(regions))
+		}
+	}
+
+	if c.AvailabilityZone == "" {
+		glog.V(4).Infof("Trying to default availability zone...")
+		availabilityZones, err := getAvailabilityZones(client, c.Region)
+		if err != nil {
+			return spec, changed, fmt.Errorf("Failed to get availability zones: '%v'", err)
+		}
+		if len(availabilityZones) == 1 {
+			glog.V(4).Infof("Defaulted availability Zone to '%s'", availabilityZones[0].ZoneName)
+			changed = true
+			c.AvailabilityZone = availabilityZones[0].ZoneName
+		}
+	}
+
+	var subnetsFromDefaultedNetwork []string
+	if c.Network == "" {
+		glog.V(4).Infof("Trying to default network...")
+		networks, err := getNetworks(client, c.Region)
+		if err != nil {
+			return spec, changed, fmt.Errorf("Failed to retrieve networks: '%v'", err)
+		}
+		if len(networks) == 1 {
+			glog.V(4).Infof("Defaulted network to '%s'", networks[0].Name)
+			changed = true
+			c.Network = networks[0].Name
+			c.NetworkID = networks[0].ID
+			subnetsFromDefaultedNetwork = networks[0].Subnets
+		} else {
+			// Networks without subnets can't be used, try finding a default by excluding them
+			// However the network object itself still contains the subnet, the only difference
+			// is that the subnet can not be retrieved by itself
+			var candidates []osnetworks.Network
+		NetworkLoop:
+			for _, network := range networks {
+				for _, subnet := range network.Subnets {
+					_, err := getSubnet(client, c.Region, subnet)
+					if err == errNotFound {
+						continue
+					} else if err != nil {
+						return spec, changed, fmt.Errorf("Failed to retrieve subnet: '%v'", err)
+					}
+					candidates = append(candidates, network)
+					continue NetworkLoop
```

Comment:
Can we put the logic for retrieving the default network into a separate function?